### ISSUE TITLE
[miniupnpd] iptables_init nat chain is ignored under common conditions

### ISSUE THREAD

**Issue by 5paceToast** [miniupnp/miniupnp#354]
The current implementation of `iptables_init.sh` will add the nat rule at the end of the POSTROUTING chain.
In most gateways, there is already a similar (less specific) rule that applies catch-all MASQUERADE to the same condition (`-o external_interface`).
As such, the current miniupnpd nat table will never be used, as MASQUERADE is a terminating target.

This was mentioned in a [comment](https://github.com/miniupnp/miniupnp/commit/c3f752db4a8286be00ad1d8b2a9fab37dc8d116d#r32668594) in c3f752d (which was added to solve #334), and as per the [response](https://github.com/miniupnp/miniupnp/commit/c3f752db4a8286be00ad1d8b2a9fab37dc8d116d#r32673717) am opening another issue to address that.

**Comment by miniupnp** [comment]
we could add an option to choose between -A and -I.

Have you a specific example of where it will be useful ?

**Comment by 5paceToast** [comment]
If you already have a firewall (common on gateways - the target of miniupnpd), and already have a generic masquerade line (`-o $EXTIF -j MASQUERADE`), miniupnpd's entrance will NEVER get called.
The useful use-case is having miniupnpd actually work as intended.

On the other hand, I do not see any issues with using -I.
After all, it *only* affects entries that are supposed to be forwarded, and passes through otherwise.

This means that for someone to *not* want -I, they must want to override dynamically generated rules, which suggests they should modify the configuration file, rather than touch the firewall.

The same applies to PREROUTING, though that's generally less commonly touched, and there's no commonly deployed rule that makes it not work.

tl;dr
- Using -I does not break anything.
- Using -A (silently) breaks miniupnpd on the vast majority of intended platforms.

**Comment by miniupnp** [comment]
miniupnpd has used -A from the start (2006...) and I don't remember someone never complained until now...

can you be more explicit about what are " the vast majority of intended platforms" ?
I'm curious to see the full tables of your configuration (and what gateway it is ?)

**Comment by 5paceToast** [comment]
"the vast majority of intended platforms" will have NAT.
NAT is most commonly implemented by having the following iptables rule:
`iptables -t nat -A POSTROUTING -o eth0 -j MASQUERADE`.
If miniupnpd is loaded after this rule, its rule will never be reached (since it'll never get there, due to having a subset of the condition, and MASQUERADE being a terminating rule).

As for my setup...
My gateway is Linux 4.19.21 running Alpine Edge.
My firewall is written in ferm, here is the entirety of it:
```
@def $DEV_WAN = ( doesn't matter );
@def $DEV_LAN = ( doesn't matter );

@def $NET_LAN = doesn't matter;
@def $NET_WG = doesn't matter;
@def $NET_LOCAL = ( $NET_LAN $NET_WG );

# globally accessible services
@def $TCP_WAN = ( 22 80 443 );
@def $UDP_WAN = ( 1194 );
# ( ssh http https ) ( wireguard )

# locally accessible services
@def $TCP_LAN = ( 53 19080 );
@def $UDP_LAN = ( 53 67 69 123 547 1900 5351 );
# ( dns upnp_http ) ( dns dhcp tftp ntp dhcp6 upnp nat-pmp )

# local/forward services
@def $TCP_FWD = ( 22 80 443 8000 8080 );
@def $UDP_FWD = ();
# ( ssh 80 443 laptop_edit1 laptop_edit2 )

# shared
domain (ip ip6) {
    table filter {
        chain (INPUT FORWARD) {
            policy DROP;

            mod conntrack ctstate (RELATED ESTABLISHED) ACCEPT;
            mod conntrack ctstate INVALID DROP;

            proto icmp ACCEPT;
        }
        chain INPUT {
            interface lo ACCEPT;

            interface $DEV_LAN {
                proto tcp dport $TCP_LAN ACCEPT;
                proto udp mod multiport destination-ports $UDP_LAN ACCEPT;
            }

            proto tcp dport $TCP_WAN ACCEPT;
            proto udp dport $UDP_WAN ACCEPT;
        }
        chain FORWARD {
            interface $DEV_LAN ACCEPT;

            proto tcp dport $TCP_FWD ACCEPT;
            proto udp dport $UDP_FWD ACCEPT;

            mod conntrack ctstate DNAT ACCEPT;
        }
    }
}

table nat chain POSTROUTING {
    policy ACCEPT;
    outerface $DEV_WAN MASQUERADE;
    saddr $NET_LOCAL mod conntrack ctstate DNAT MASQUERADE; # needle point loopback
}

table nat chain PREROUTING policy ACCEPT;
```
The firewall is habitually loaded *before* services such as miniupnpd.
Here is the dump of `iptables -t nat -S POSTROUTING`:
```
-P POSTROUTING ACCEPT
-A POSTROUTING -o iface0 -j MASQUERADE
-A POSTROUTING -o iface1 -j MASQUERADE
-A POSTROUTING -s prefix/X -m conntrack --ctstate DNAT -j MASQUERADE
-A POSTROUTING -s prefix/X -m conntrack --ctstate DNAT -j MASQUERADE
-A POSTROUTING -o iface0 -j MINIUPNPD-POSTROUTING
```

As you see, there is no way MINIUPNPD-POSTROUTING would *ever* be reached.
On the other hand, it has no way of interfering with anything (unless it *should*, since there's a upnp rule in play).

**Comment by miniupnp** [comment]
many user may well load the other rules AFTER miniupnpd.

I'm wondering, should we offer to choose -A or -I for each table independently or globally ?

**Comment by 5paceToast** [comment]
Various programs that modify netfilter are habitually loaded after the primary rules.
This is because most ways to add the primary table rules will flush whatever is in your tables right now (for example, see iptables-restore(8)).

In order to load rules *after* loading miniupnpd, one would require to break common linux conventions, and modify every related service script/file/unit (for example to use `-n` with iptables-restore, causing potential further side-effects).

For an example of a different application that modifies netfilter tables, see docker:
It uses -I in since it knows there's no chance of (legitimate) interference (as the rules are carefully crafted to let all non-managed traffic through, as in miniupnpd).

As for whether or not it's ok to use -I in all tables, it depends on what kind of rules miniupnpd will generate.
So long as 100% of the terminating rules miniupnpd generates are directly related to desired behavior (e.g letting known upnp traffic through), there's really no reason to just use -I globally.

**Comment by miniupnp** [comment]
@5paceToast please tell me if 476974ab52ffa09509fe9ce9ab6f577b4d0e3d63 is OK for you

(branch https://github.com/miniupnp/miniupnp/tree/issue_354)

**Comment by 5paceToast** [comment]
@miniupnp looks good!
I successfully established a test connection (using upnpc and an external server).
I'm glad we could get a fix in upstream - this way all users can benefit :)
